vowpalwabbit/parse_primitives.cc

```diff
@@ -53,6 +53,82 @@ std::vector<std::string> escaped_tokenize(char delim, VW::string_view s, bool al
   return tokens;
 }
 
+bool is_delim(char c) { return c == ' '; }
+
+bool is_quote(char c) { return c == '"' || c == '\''; }
+
+bool is_escape_char(char c) { return c == '\\'; }
+
+template <typename It>
+char unescape_char(It char_to_unescape_it, It end)
+{
+  if (char_to_unescape_it == end) { THROW("unescape_char: unexpected end of string while unescaping"); }
+
+  char c = *char_to_unescape_it;
+
+  if (c == 'n') { return '\n'; }
+
+  return c;
+}
+
+template <typename It>
+std::vector<std::string> split_impl(It begin, It end)
+{
+  std::vector<std::string> ret;
+  if (begin == end) { return ret; }
+
+  bool inside_quote = false;
+  char quote_char = '\0';
+  std::string current;
+  for (; begin != end; ++begin)
+  {
+    if (is_escape_char(*begin))
+    {
+      begin++;
+      current.append(1, unescape_char(begin, end));
+    }
+    else if (is_delim(*begin))
+    {
+      // If we're not inside a token, this token is done. Otherwise just add the space to the token.
+      if (!inside_quote)
+      {
+        if (!current.empty()) { ret.push_back(current); }
+        current.clear();
+      }
+      else
+      {
+        current.append(1, *begin);
+      }
+    }
+    else if (is_quote(*begin))
+    {
+      if (inside_quote && quote_char == *begin)
+      {
+        inside_quote = !inside_quote;
+        quote_char = '\0';
```

Comment:
Not sure this line is needed

Reply:
It is. `quote_char` is used to match the corresponding quote `"` or `'`. \0 is not in that range and used as a special value to make the above conditional work. This is what allows sequences like this to work `"'"`

Reply:
Right so quote_char is only relevant when inside_quote == true. Since inside_quote is set to false here, quote_char won't be relevant again until it's reset in the next conditional.

Reply:
Oh you're so right. Good catch!